### ISSUE TITLE
In Cluster Deployment RBAC issue

### ISSUE THREAD

**Issue by utkarsh-pro** [meshery/meshery#3881]
#### Current Behavior
Running `mesheryctl system start -p kubernetes` deploys meshery in k8s cluster. Meshery has been using custom `ServiceAccount` with appropriate role and role bindings for a while now but still in cluster deployments are using the `default` service acccount which causes Authorization issues.

Upon investigating the issue, the issue seems to be coming from `mesheryctl`. 
```
$ mesheryctl system start -p kubernetes
...
downloading manifest files from https://github.com/meshery/meshery/tree/master/install/deployment_yamls/k8s
```
Although the logs says that the manifests are being downloaded from `master` yet the manifests in the `~/.meshery/manifests` does not matches with the `master` branch manifests.

#### Expected Behavior
Meshery should be deployed with appropriate service account. It should also be ensured that service account is created first and then meshery is deployed.

#### Screenshots/Logs
![image](https://user-images.githubusercontent.com/45818886/129449707-453cd99f-3655-40a7-bcb6-44cba781e50a.png)

#### Environment

- **Host OS:** Linux
- **Platform:** Kubernetes
- **Meshery Server Version:** edge-latest
- **Meshery Client Version:** stable-v0.5.46 

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and Resources
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

**Comment by Revolyssup** [comment]
@utkarsh-pro  can you also share the config.yaml

**Comment by leecalcote** [comment]
+ amitech

**Comment by leecalcote** [comment]
\+ @ amitech

**Comment by amitech** [comment]
```yaml
contexts:
  local:
    endpoint: http://192.168.39.221:30477
    token: Default
    platform: kubernetes
    adapters:
    - meshery-istio
    - meshery-linkerd
    - meshery-consul
    - meshery-nsm
    - meshery-kuma
    - meshery-cpx
    - meshery-osm
    - meshery-traefik-mesh
    channel: stable
    version: latest
  test:
    adapters:
    - meshery-istio
    - meshery-linkerd
    - meshery-consul
    - meshery-nsm
    - meshery-kuma
    - meshery-cpx
    - meshery-osm
    - meshery-traefik-mesh
    channel: stable
    endpoint: http://localhost:9081
    platform: docker
    token: Default
    version: latest
current-context: local
tokens:
- location: auth.json
  name: Default
```

**Comment by leecalcote** [comment]
@metonymic-smokey fyi

**Comment by leecalcote** [comment]
@samirparhi-dev, is this one that you might address?

**Comment by samirparhi-dev** [comment]
Yes , I will look into it.

**Comment by metonymic-smokey** [comment]
Samir,
If you need some help with/someone to work on the mesheryctl part of it let
me know! This issue is pretty interesting.

**Comment by samirparhi-dev** [comment]
@metonymic-smokey : once i fix the service account issue , i will let you know to check it from `mctl` side. But I am glad that you jumped in :-)

**Comment by samirparhi-dev** [comment]
@metonymic-smokey : Can you help me understand how the manifest is getting downloaded with `mesheryctl system start` command from the code perspective ?

**Comment by metonymic-smokey** [comment]
Hey Samir,
So for platform K8s, first `FetchManifests` is invoked. Based on the channel and version in the current context, `GetManifestTreeURL` is invoked and returns a manifests URL for that version, for manifests starting with `install/deployment_yamls/k8s`.  
Then a manifests folder is created(`~/.meshery/manifests`).
Finally, `DownloadManifests` downloads manifest files into that folder for all the file names passed to it. The rawURL passed to it is customised for each file name.

**Comment by metonymic-smokey** [comment]
@utkarsh-pro what difference did you find in the downloaded manifests and the ones on the master branch?

**Comment by samirparhi-dev** [comment]
@metonymic-smokey  :A per my testing , it is not downloading the `desired manifest` form the desired `channel` and `release`

**Comment by metonymic-smokey** [comment]
What is the difference you're observing?

**Comment by utkarsh-pro** [comment]
@metonymic-smokey downloaded manifests doesn't have `serviceAccountName` in the pod spec of the deployments.

**Comment by metonymic-smokey** [comment]
Thanks!
https://raw.githubusercontent.com/meshery/meshery/master/install/deployment_yamls/k8s/meshery-deployment.yaml - this is where the manifests are downloaded. This has the `serviceAccountParameter` so I don't think the bug is in the downloading.

**Comment by tangledbytes** [comment]
@metonymic-smokey this is the meshery-deployment.yaml manifest that is present in my `~/.meshery/manifests` directory

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    kompose.cmd: kompose convert -f ../docker-compose.yaml
    kompose.version: 1.17.0 ()
  creationTimestamp: null
  labels:
    io.kompose.service: meshery
  name: meshery
spec:
  replicas: 1
  selector:
    matchLabels:
      io.kompose.service: meshery
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        io.kompose.service: meshery
    spec:
      containers:
      - env:
        - name: EVENT
          value: mesheryLocal
        - name: PROVIDER_BASE_URLS
          value: https://meshery.meshery.io
        - name: ADAPTER_URLS
          value: meshery-istio:10000 meshery-linkerd:10001 meshery-consul:10002 meshery-octarine:10003
            meshery-nsm:10004 meshery-kuma:10007 meshery-cpx:10008 meshery-osm:10009
            meshery-nginx-sm:10010
        image: meshery/meshery:edge-latest
        name: meshery
        ports:
        - containerPort: 8080
        resources: {}
      restartPolicy: Always
status: {}
```

This manifest **does not** have `serviceAccountName` while the link that you have pointed above does. This field is extremely crucial.

**Comment by utkarsh-pro** [comment]
@metonymic-smokey I believe this is where the bug is: https://github.com/meshery/meshery/blob/master/mesheryctl/pkg/utils/compose.go#L24

Instead of `serviceAccount` it should be `serviceAccountName`

What do you think?

**Comment by metonymic-smokey** [comment]
@utkarsh-pro got your point, thanks!
I put a `time.Sleep` in ChangingManifestVersion and interrupted it immediately after the download. This time, my manifest file has `serviceAccount` in the spec.
So I think it might have something to do with that.

**Comment by metonymic-smokey** [comment]
I think you narrowed down the cause correctly. Since the mapping does not match, that field is not parsed when unmarshalled and hence not written to the manifests.

**Comment by metonymic-smokey** [comment]
Just tried it locally, your fix works!

**Comment by samirparhi-dev** [comment]
> @metonymic-smokey I believe this is where the bug is: https://github.com/meshery/meshery/blob/master/mesheryctl/pkg/utils/compose.go#L24
> 
> Instead of `serviceAccount` it should be `serviceAccountName`
> 
> What do you think?

Yes that is Correct `serviceAccount` it should be `serviceAccountName`

**Comment by leecalcote** [comment]
Why is the manifest being changed after download?

Sleeping is nearly always not the right answer. 

Is there an issue open to move away from manifests to helm?

**Comment by utkarsh-pro** [comment]
@leecalcote, @metonymic-smokey placed `time.Sleep` to debug the downloaded file. This allowed her to interrupt mesheryctl just after the downloading finished, I don't think this was proposed as a solution.

Manifest is being changed because container image names have version defined and these have to be changed as per the channel. 

I don't think we have an issue open for that just yet.

**Comment by leecalcote** [comment]
Thank you much, @utkarsh-pro. Makes sense. 

@metonymic-smokey One thing that I believe we have a bug lurking around is the undoing of channel subscriptions. If some is subscribed to "latest", their manifests need to reflect this "latest" subscription, and not be pinned to a version. 

@metonymic-smokey will you kindly open an issue for transitioning to helm?

**Comment by metonymic-smokey** [comment]
@leecalcote #3291 is an open issue about transition to Helm.

**Comment by metonymic-smokey** [comment]
I think this issue can be closed since #3959 had a commit fixing it.

**Comment by samirparhi-dev** [comment]
closing the issue
